dll/win32/shell32/folders/CDrivesFolder.cpp

```diff
@@ -57,6 +58,85 @@ static int iDriveTypeIds[7] = { IDS_DRIVE_FIXED,       /* DRIVE_UNKNOWN */
 *   IShellFolder implementation
 */
 
+static BOOL TryToLockDrive(HANDLE hDrive, DWORD dwRetryCount, DWORD dwSleep)
+{
+    DWORD dwError, dwBytesReturned;
+    for (DWORD i = 0; i < dwRetryCount; ++i)
+    {
+        if (DeviceIoControl(hDrive, FSCTL_LOCK_VOLUME, NULL, 0, NULL, 0, &dwBytesReturned, NULL))
+            return TRUE;
+
+        dwError = GetLastError();
+        if (dwError == ERROR_INVALID_FUNCTION)
+            break; /* don't sleep if function is not implemented */
+
+        Sleep(dwSleep);
+    }
+    SetLastError(dwError);
+    return FALSE;
+}
+
+// NOTE: See also https://support.microsoft.com/en-us/help/165721/how-to-ejecting-removable-media-in-windows-nt-windows-2000-windows-xp
+static BOOL DoEjectDrive(const WCHAR *physical, UINT nDriveType, INT& nStringID)
+{
+    DWORD dwAccessMode, dwShareMode = FILE_SHARE_READ | FILE_SHARE_WRITE;
+    if (nDriveType == DRIVE_REMOVABLE)
+        dwAccessMode = GENERIC_READ | GENERIC_WRITE;
+    else if (nDriveType == DRIVE_CDROM)
+        dwAccessMode = GENERIC_READ;
+
+    HANDLE hDrive = CreateFile(physical, dwAccessMode, dwShareMode, 0, OPEN_EXISTING, 0, NULL);
+    if (hDrive == INVALID_HANDLE_VALUE)
+        return FALSE;
+
+    BOOL bResult;
+    DWORD dwBytesReturned, dwError = NO_ERROR;
+    do
+    {
+        PREVENT_MEDIA_REMOVAL removal;
+        bResult = TryToLockDrive(hDrive, 3, 250);
+        if (!bResult)
+        {
+            dwError = GetLastError();
+            if (dwError != ERROR_INVALID_FUNCTION)  /* temporary hack */
+            {
+                nStringID = IDS_CANTLOCKVOLUME; /* Unable to lock volume */
+                break;
+            }
+        }
+        bResult = DeviceIoControl(hDrive, FSCTL_DISMOUNT_VOLUME, NULL, 0, NULL, 0, &dwBytesReturned, NULL);
+        if (!bResult)
+        {
+            dwError = GetLastError();
+            if (dwError != ERROR_INVALID_FUNCTION)  /* temporary hack */
```

Comment:
Why?!

Reply:
@binarymaster says so.

Reply:
I adviced that since you said ejection works without these checks.

But @HeisSpiter knows better.

Reply:
Ejection might work without locking. Actually, IIRC, MS FastFAT allows lock-free dismounts. But it still implements volume un/locking. So, I'd rather we do it properly: we always try to lock, even if pointless for some FSD.
A FSD not implementing the primitives for un/locking looks rather suspicious (see ROS one ;-)) and the error should not be ignored.